### PR TITLE
Decayed reservoir sampling

### PR DESCRIPTION
**Pull request by hnyu**
This doesn't support RNN states (i.e., sampling a temporal sequence) because of the nature of its sampling strategy.

### CODE REVIEW

File: alf/utils/reservoir_sampler.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      (s+1)*K/t * (t+1-s-1)/(t+1) * (t+2-s-1)/(t+2) * ... * (T-s-1)/T
+
+      which is proportional to (t-1)(t-2)..(t-s) / ((T-1)(T-2)...(T-s)) and is
+      roughly proportional to t^s, when t,T >> (s+1)*K
+    """
+
+    def __init__(self,
+                 data_spec: tf.TensorSpec,
+                 K,
+                 speed=0,
```
Comment (emailweixu) [inline alf/utils/reservoir_sampler.py]: Feel we should make speed mean the same thing as the speed for AdaptiveAverager and AdaptiveNormalizer

File: alf/utils/reservoir_sampler.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        buffer_space_left = self._capacity - self._current_size
+        batch_replacing_size = batch_size - buffer_space_left
+
+        # Split the batch into two: the first one (if exists) will be directly
+        # added into the buffer; the second one (if exists) will randomly replace
+        # samples in the buffer
+        def _fill_buffer():
+            data_filling_buffer = tf.nest.map_structure(
+                lambda bat: bat[:buffer_space_left], batch)
+            super(ReservoirSampler, self).add_batch(data_filling_buffer)
```
Comment (emailweixu) [inline alf/utils/reservoir_sampler.py]: super().add_batch

File: alf/utils/reservoir_sampler.py (hunk 1 above)
Comment (hnyu) [reply alf/utils/reservoir_sampler.py]: changed

File: alf/utils/reservoir_sampler.py (hunk 2 above)
Comment (hnyu) [reply alf/utils/reservoir_sampler.py]: That will report errors about incorrect arguments for super().

File: alf/utils/reservoir_sampler.py [hunk 3]
```diff
@@ -28,29 +28,29 @@ class ReservoirSampler(DataBuffer):
...
+      t+m is (1 - s*K/(t+m) * 1/K) = (t+m-s) / (t+m)
     - The over probability of the t-th sample still being kept in the set at time
       T is
 
-      (s+1)*K/t * (t+1-s-1)/(t+1) * (t+2-s-1)/(t+2) * ... * (T-s-1)/T
+      s*K/t * (t+1-s)/(t+1) * (t+2-s)/(t+2) * ... * (T-s)/T
 
-      which is proportional to (t-1)(t-2)..(t-s) / ((T-1)(T-2)...(T-s)) and is
-      roughly proportional to t^s, when t,T >> (s+1)*K
+      which is proportional to t(t-1)..(t-s+1) / ((T-1)(T-2)...(T-s+1)) and is
```
Comment (emailweixu) [inline alf/utils/reservoir_sampler.py]: t(t-1)..(t-s+1) => (t-1)..(t-s+1)

File: alf/utils/reservoir_sampler.py [hunk 4]
```diff
@@ -59,13 +59,13 @@ def __init__(self,
...
             speed (int): a bigger speed results in a faster decay of old samples.
-                When `speed` is 0, it results in the original reservoir sampling.
+                When `speed` is 1, it results in the original reservoir sampling.
             name (str): name of the sampler
         """
         super(ReservoirSampler, self).__init__(
             data_spec=data_spec, capacity=K, name=name)
         assert isinstance(K, int) and K > 0
-        assert isinstance(speed, int) and speed >= 0
+        assert isinstance(speed, int) and speed > 0
```
Comment (emailweixu) [inline alf/utils/reservoir_sampler.py]: I think speed doesn't have to be integer as long as it is greater than or equal to one

File: alf/utils/reservoir_sampler.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+      s*K/t * (t+1-s)/(t+1) * (t+2-s)/(t+2) * ... * (T-s)/T
+
+      which is proportional to t(t-1)..(t-s+1) / ((T-1)(T-2)...(T-s+1)) and is
+      roughly proportional to t^(s-1), when t,T >> s*K
+    """
+
+    def __init__(self,
+                 data_spec: tf.TensorSpec,
+                 K,
```
Comment (emailweixu) [inline alf/utils/reservoir_sampler.py]: `capacity` is easier to understand

**Pull request merged**